File: src/isalhg/core/structural_tuples.py
"""Structural tuples ``xi`` (per-node) and ``eta`` (per-edge).

``xi_h(v)`` = number of primal-graph neighbours of ``v`` at distance exactly ``h``.
``eta_h(e)`` = sum over ``v in e`` of ``xi_h(v)``.

Used both for seed selection (argmax-lex ``xi`` over nodes, invariant 4) and
for tie-breaking during greedy H2S (lex ``eta`` over candidate edges).
Default depth ``3`` is inherited from IsalGraph (invariant 8); deviation
requires re-validating Theorem 2 (canonical completeness conjecture).

Both implementations live in this module: the pure-Python reference under
``_python_max_xi_nodes`` and the C++-backed wrapper under
``_cpp_max_xi_nodes``. :func:`max_xi_nodes` dispatches between them via
the ``backend=`` kwarg (default ``"cpp"``).
"""

from __future__ import annotations

from isalhg.core._core import (
    max_neighbor_degree_nodes as _core_max_neighbor_degree_nodes,
)
from isalhg.core._core import (
    max_xi_nodes as _core_max_xi_nodes,
)
from isalhg.core.backends import Backend, resolve
from isalhg.core.sparse_hypergraph import SparseHypergraph
from isalhg.types import EdgeId, NodeId

DEFAULT_DEPTH: int = 3


def _bfs_counts(H: SparseHypergraph, v: NodeId, depth: int) -> list[set[NodeId]]:
    """Return ``[shell_1, ..., shell_depth]`` where ``shell_h`` is the set of
    vertices at primal-graph distance exactly ``h`` from ``v``.
    """
    adj = H.primal_graph()
    seen: set[NodeId] = {v}
    frontier: set[NodeId] = {v}
    shells: list[set[NodeId]] = []
    for _ in range(depth):
        nxt: set[NodeId] = set()
        for u in frontier:
            nxt.update(adj[u] - seen)
        seen.update(nxt)
        shells.append(nxt)
        frontier = nxt
    return shells


def xi(H: SparseHypergraph, v: NodeId, depth: int = DEFAULT_DEPTH) -> tuple[int, ...]:
    """Return ``(|shell_1(v)|, ..., |shell_depth(v)|)``."""
    if depth < 1:
        return ()
    shells = _bfs_counts(H, v, depth)
    return tuple(len(s) for s in shells)


def xi_labelled(
    H: SparseHypergraph,
    v: NodeId,
    depth: int = DEFAULT_DEPTH,
) -> tuple[tuple[int, ...], ...]:
    """Label-aware xi: ``[(count_per_label at h=1), (... at h=2), ...]``.

    For the trivial vocabulary this is equivalent to ``xi`` lifted into a
    1-element inner tuple per shell.
    """
    if depth < 1:
        return ()
    shells = _bfs_counts(H, v, depth)
    sigma_v = H.n_vertex_labels
    out: list[tuple[int, ...]] = []
    for shell in shells:
        counts = [0] * sigma_v
        for u in shell:
            counts[H.vertex_label(u)] += 1
        out.append(tuple(counts))
    return tuple(out)


def eta(H: SparseHypergraph, e: EdgeId, depth: int = DEFAULT_DEPTH) -> tuple[int, ...]:
    """Return ``(eta_1(e), ..., eta_depth(e))`` where
    ``eta_h(e) = sum_{v in e} xi_h(v)``.
    """
    if depth < 1:
        return ()
    accum = [0] * depth
    for v in H.members(e):
        for h, count in enumerate(xi(H, v, depth)):
            accum[h] += count
    return tuple(accum)


def _seed_key(H: SparseHypergraph, v: NodeId, depth: int) -> tuple[object, ...]:
    """Composite seed-selection key.

    Maximised lexicographically:
    1. ``xi_labelled(v)`` -- shell-by-shell label-aware count tuple.
    2. ``H.vertex_label(v)`` -- the seed's own label.
    """
    return (xi_labelled(H, v, depth), H.vertex_label(v))


def _python_max_xi_nodes(
    H: SparseHypergraph,
    depth: int = DEFAULT_DEPTH,
) -> tuple[NodeId, ...]:
    """Pure-Python reference implementation of :func:`max_xi_nodes`."""
    if H.n_nodes == 0:
        return ()
    best_key = max(_seed_key(H, v, depth) for v in H.nodes())
    return tuple(v for v in H.nodes() if _seed_key(H, v, depth) == best_key)


def _cpp_max_xi_nodes(
    H: SparseHypergraph,
    depth: int = DEFAULT_DEPTH,
) -> tuple[NodeId, ...]:
    """C++-backed implementation of :func:`max_xi_nodes`."""
    if H.n_nodes == 0:
        return ()
    return tuple(_core_max_xi_nodes(H, depth))


_MAX_XI_BACKENDS: dict[str, object] = {
    "python": _python_max_xi_nodes,
    "cpp": _cpp_max_xi_nodes,
}


def max_xi_nodes(
    H: SparseHypergraph,
    depth: int = DEFAULT_DEPTH,
    *,
    backend: Backend | None = None,
) -> tuple[NodeId, ...]:
    """Return all nodes attaining the lexicographic maximum of the seed key.

    Invariant 4: one iso-invariant seed set for the canonical algorithm
    (the historical ``xi`` cascade). Any iso-invariant node set is
    admissible; :func:`max_neighbor_degree_nodes` is the tighter default
    used since T-M0. The soundness of the canonical string requires only
    that the seed set is preserved by hypergraph automorphisms.

    Parameters
    ----------
    H, depth
        See module docstring.
    backend : {"cpp", "python"}, optional
        Implementation to use. Defaults to ``"cpp"`` (see
        :data:`isalhg.core.backends.DEFAULT_BACKEND`).
    """
    if H.n_nodes == 0:
        return ()
    impl = resolve(backend, _MAX_XI_BACKENDS)
    return impl(H, depth)


# ---------------------------------------------------------------------------
# PI 2026-06-23 — neighbour-degree seed selector.
#
# Cascade: max vertex label → max primal-graph degree → lex-max
# (sorted-descending list of neighbour primal-graph degrees). Each rung is
# an iso-invariant projection of the vertex set, so the intersection is
# iso-invariant. Strictly cheaper than max_xi_nodes (no depth-3 BFS) and
# typically more discriminating on non-vertex-transitive inputs.
# ---------------------------------------------------------------------------


def _neighbour_degree_key(adj: dict[NodeId, set[NodeId]], v: NodeId) -> tuple[int, ...]:
    """Descending-sorted list of primal-graph degrees of ``v``'s neighbours.

    Parameters
    ----------
    adj : dict[NodeId, set[NodeId]]
        Pre-built primal-graph adjacency (caller's responsibility to build once).
    v : NodeId
        Target node.
    """
    return tuple(sorted((len(adj[u]) for u in adj[v]), reverse=True))


def _python_max_neighbor_degree_nodes(
    H: SparseHypergraph,
) -> tuple[NodeId, ...]:
    """Pure-Python reference implementation of :func:`max_neighbor_degree_nodes`."""
    if H.n_nodes == 0:
        return ()
    adj = H.primal_graph()

    # Step 1 — max vertex label.
    max_label = max(H.vertex_label(v) for v in H.nodes())

    # Step 2 — max primal-graph degree among label-max vertices.
    survivors = [v for v in H.nodes() if H.vertex_label(v) == max_label]
    max_deg = max(len(adj[v]) for v in survivors)
    survivors = [v for v in survivors if len(adj[v]) == max_deg]

    # Step 3 — lex-max sorted-desc neighbour-degree list.
    keys = {v: _neighbour_degree_key(adj, v) for v in survivors}
    best_key = max(keys.values())
    return tuple(v for v in survivors if keys[v] == best_key)


def _cpp_max_neighbor_degree_nodes(
    H: SparseHypergraph,
) -> tuple[NodeId, ...]:
    """C++-backed implementation of :func:`max_neighbor_degree_nodes`."""
    if H.n_nodes == 0:
        return ()
    return tuple(_core_max_neighbor_degree_nodes(H))


_MAX_NBR_DEG_BACKENDS: dict[str, object] = {
    "python": _python_max_neighbor_degree_nodes,
    "cpp": _cpp_max_neighbor_degree_nodes,
}


def max_neighbor_degree_nodes(
    H: SparseHypergraph,
    *,
    backend: Backend | None = None,
) -> tuple[NodeId, ...]:
    """Return the PI 2026-06-23 lex-max neighbour-degree seed set.

    Cascade: max vertex label → max primal-graph degree → lex-max
    descending-sorted list of neighbour primal-graph degrees. Iso-invariant
    by construction (each rung is preserved by hypergraph automorphisms),
    so the resulting seed set is a sound replacement for
    :func:`max_xi_nodes` inside the canonical-string lex-min loop.

    Parameters
    ----------
    H : SparseHypergraph
        Connected hypergraph.
    backend : {"cpp", "python"}, optional
        Implementation to use. Defaults to ``"cpp"``.
    """
    if H.n_nodes == 0:
        return ()
    impl = resolve(backend, _MAX_NBR_DEG_BACKENDS)
    return impl(H)

File: src/isalhg/core/_native/src/structural_tuples.cpp
#include "isalhg/structural_tuples.hpp"

#include <algorithm>
#include <functional>

namespace isalhg {

std::vector<std::vector<std::int32_t>> xi_labelled_counts(const SHG& H, NodeId v, int depth) {
    const int sigma_v = std::max(1, H.n_vertex_labels);
    std::vector<std::vector<std::int32_t>> out(
        static_cast<std::size_t>(std::max(0, depth)),
        std::vector<std::int32_t>(static_cast<std::size_t>(sigma_v), 0));
    if (depth < 1 || H.n_nodes == 0) return out;

    std::vector<std::uint8_t> seen(static_cast<std::size_t>(H.n_nodes), 0);
    std::vector<NodeId> frontier;
    std::vector<NodeId> next_frontier;
    frontier.reserve(static_cast<std::size_t>(H.n_nodes));
    next_frontier.reserve(static_cast<std::size_t>(H.n_nodes));

    seen[static_cast<std::size_t>(v)] = 1;
    frontier.push_back(v);

    for (int h = 0; h < depth; ++h) {
        next_frontier.clear();
        for (NodeId u : frontier) {
            for (NodeId w : H.primal_adj[static_cast<std::size_t>(u)]) {
                if (!seen[static_cast<std::size_t>(w)]) {
                    seen[static_cast<std::size_t>(w)] = 1;
                    next_frontier.push_back(w);
                }
            }
        }
        auto& counts = out[static_cast<std::size_t>(h)];
        for (NodeId w : next_frontier) {
            const int lab = H.vertex_labels[static_cast<std::size_t>(w)];
            ++counts[static_cast<std::size_t>(lab)];
        }
        frontier.swap(next_frontier);
    }
    return out;
}

namespace {

// Lex compare two label-count tuples (flattened representation).
int compare_xi_keys(const std::vector<std::vector<std::int32_t>>& a,
                    const std::vector<std::vector<std::int32_t>>& b) noexcept {
    const std::size_t common_h = std::min(a.size(), b.size());
    for (std::size_t h = 0; h < common_h; ++h) {
        const auto& ah = a[h];
        const auto& bh = b[h];
        const std::size_t common_l = std::min(ah.size(), bh.size());
        for (std::size_t l = 0; l < common_l; ++l) {
            if (ah[l] != bh[l]) return ah[l] < bh[l] ? -1 : 1;
        }
        if (ah.size() != bh.size()) return ah.size() < bh.size() ? -1 : 1;
    }
    if (a.size() != b.size()) return a.size() < b.size() ? -1 : 1;
    return 0;
}

}  // namespace

std::vector<NodeId> max_xi_nodes_compute(const SHG& H, int depth) {
    std::vector<NodeId> out;
    if (H.n_nodes == 0) return out;

    // Compute (xi_labelled(v), vertex_label(v)) per vertex.
    std::vector<std::vector<std::vector<std::int32_t>>> xi_per_v(
        static_cast<std::size_t>(H.n_nodes));
    for (NodeId v = 0; v < H.n_nodes; ++v) {
        xi_per_v[static_cast<std::size_t>(v)] = xi_labelled_counts(H, v, depth);
    }

    // Find lex-max key across (xi_labelled, vertex_label).
    NodeId best = 0;
    auto better_than_best = [&](NodeId cand) -> bool {
        const int c = compare_xi_keys(xi_per_v[static_cast<std::size_t>(cand)],
                                      xi_per_v[static_cast<std::size_t>(best)]);
        if (c != 0) return c > 0;
        return H.vertex_labels[static_cast<std::size_t>(cand)]
                   > H.vertex_labels[static_cast<std::size_t>(best)];
    };
    for (NodeId v = 1; v < H.n_nodes; ++v) {
        if (better_than_best(v)) best = v;
    }

    // Collect all vertices matching the best key.
    for (NodeId v = 0; v < H.n_nodes; ++v) {
        const int c = compare_xi_keys(xi_per_v[static_cast<std::size_t>(v)],
                                      xi_per_v[static_cast<std::size_t>(best)]);
        if (c == 0 && H.vertex_labels[static_cast<std::size_t>(v)]
                          == H.vertex_labels[static_cast<std::size_t>(best)]) {
            out.push_back(v);
        }
    }
    return out;
}

// ---------------------------------------------------------------------------
// PI 2026-06-23 seed selector.
//
// Cascade: max label → max degree → lex-max (sorted-desc neighbour-degree
// list). Each rung is an iso-invariant projection of the vertex set, so
// the intersection is an iso-invariant subset. Cheaper than
// max_xi_nodes_compute (no depth-3 BFS), and on non-vertex-transitive
// inputs typically returns a strictly smaller set.
// ---------------------------------------------------------------------------

std::vector<NodeId> max_neighbor_degree_nodes_compute(const SHG& H) {
    std::vector<NodeId> out;
    if (H.n_nodes == 0) return out;

    // Step 1 — max vertex label.
    VertexLabel max_label = H.vertex_labels[0];
    for (NodeId v = 1; v < H.n_nodes; ++v) {
        const VertexLabel l = H.vertex_labels[static_cast<std::size_t>(v)];
        if (l > max_label) max_label = l;
    }

    // Step 2 — among label-max vertices, max primal-graph degree.
    std::int32_t max_deg = -1;
    for (NodeId v = 0; v < H.n_nodes; ++v) {
        if (H.vertex_labels[static_cast<std::size_t>(v)] != max_label) continue;
        const std::int32_t d =
            static_cast<std::int32_t>(H.primal_adj[static_cast<std::size_t>(v)].size());
        if (d > max_deg) max_deg = d;
    }

    // Collect the (label-max, degree-max) survivors. ``survivors`` is built
    // in ascending node-id order so the final ``out`` is automatically sorted.
    std::vector<NodeId> survivors;
    survivors.reserve(static_cast<std::size_t>(H.n_nodes));
    for (NodeId v = 0; v < H.n_nodes; ++v) {
        if (H.vertex_labels[static_cast<std::size_t>(v)] != max_label) continue;
        const std::int32_t d =
            static_cast<std::int32_t>(H.primal_adj[static_cast<std::size_t>(v)].size());
        if (d == max_deg) survivors.push_back(v);
    }
    if (survivors.empty()) return out;

    // Step 3 — lex-max of sorted-descending neighbour-degree list.
    auto neighbour_degree_key = [&](NodeId v) {
        const auto& nbrs = H.primal_adj[static_cast<std::size_t>(v)];
        std::vector<std::int32_t> key;
        key.reserve(nbrs.size());
        for (NodeId u : nbrs) {
            key.push_back(
                static_cast<std::int32_t>(H.primal_adj[static_cast<std::size_t>(u)].size()));
        }
        std::sort(key.begin(), key.end(), std::greater<std::int32_t>{});
        return key;
    };

    std::vector<std::vector<std::int32_t>> keys;
    keys.reserve(survivors.size());
    for (NodeId v : survivors) keys.push_back(neighbour_degree_key(v));

    auto cmp_lex = [](const std::vector<std::int32_t>& a,
                      const std::vector<std::int32_t>& b) -> int {
        const std::size_t common = std::min(a.size(), b.size());
        for (std::size_t k = 0; k < common; ++k) {
            if (a[k] != b[k]) return a[k] < b[k] ? -1 : 1;
        }
        if (a.size() != b.size()) return a.size() < b.size() ? -1 : 1;
        return 0;
    };

    std::size_t best_i = 0;
    for (std::size_t i = 1; i < survivors.size(); ++i) {
        if (cmp_lex(keys[i], keys[best_i]) > 0) best_i = i;
    }
    for (std::size_t i = 0; i < survivors.size(); ++i) {
        if (cmp_lex(keys[i], keys[best_i]) == 0) {
            out.push_back(survivors[i]);
        }
    }
    return out;
}

}  // namespace isalhg
